File: fundamentals_pipeline/60_streamlit_app/lib/format.py
"""Number formatting helpers shared across renderers."""

from __future__ import annotations

import math
from typing import Optional

import pandas as pd

EM_DASH = "—"


def is_missing(v) -> bool:
    return v is None or (isinstance(v, float) and math.isnan(v)) or pd.isna(v)


def fmt_num(v) -> str:
    """Table number cells: thousands-separated integer, accounting-style negatives."""
    if is_missing(v):
        return EM_DASH
    if v < 0:
        return f"({abs(v):,.0f})"
    return f"{v:,.0f}"


def fmt_num_scaled(v, divisor: int = 1, decimals: int | None = None) -> str:
    """Statement-table cells with an optional unit scale.

    divisor: 1 / 1e3 / 1e6 / 1e9. When divisor == 1, behaves exactly like
    fmt_num (thousands-separated integer, accounting-style negatives) so the
    'Units' view is byte-identical to today. When scaled, defaults to 1 decimal
    so small magnitudes don't collapse to 0 / round away signal (e.g. $391.0B
    rather than $391B; a $40M line under 'Billions' shows 0.0 rather than 0).
    """
    if is_missing(v):
        return EM_DASH
    if divisor == 1:
        return fmt_num(v)                       # unchanged Units path
    scaled = v / divisor
    dp = 1 if decimals is None else decimals
    if scaled < 0:
        return f"({abs(scaled):,.{dp}f})"
    return f"{scaled:,.{dp}f}"


def fmt_eps(v) -> str:
    """Per-share rows: 2 decimals, accounting-style negatives."""
    if is_missing(v):
        return EM_DASH
    if v < 0:
        return f"({abs(v):,.2f})"
    return f"{v:,.2f}"


def fmt_kpi(v) -> str:
    """KPI strip: $391.0B / $93.7M (scale-aware, one decimal)."""
    if is_missing(v):
        return EM_DASH
    sign = "-" if v < 0 else ""
    a = abs(v)
    # The export keeps raw USD. We display in $B if ≥ $1B else $M.
    # Source values are already in raw USD (not millions); financials are large numbers.
    if a >= 1_000_000_000:
        return f"{sign}${a / 1_000_000_000:.1f}B"
    if a >= 1_000_000:
        return f"{sign}${a / 1_000_000:.1f}M"
    return f"{sign}${a:,.0f}"


def fmt_metric(v, unit: Optional[str], signed: bool = False) -> str:
    """Derived metrics: format depending on unit (percent / ratio / usd).

    `signed=True` forces a leading "+" on positive percents — only for metrics
    that represent *direction* (growth YoY, margin of safety), not level
    (margins, returns, yields). Negatives always show "−" either way.
    """
    if is_missing(v):
        return EM_DASH
    if unit == "percent":
        if signed:
            return f"{v:+.1f}%"
        return f"{v:.1f}%"
    if unit == "ratio":
        return f"{v:,.2f}x"
    if unit == "usd":
        return fmt_kpi(v)
    return f"{v:,.2f}"


def fmt_mos(v) -> str:
    """Margin-of-safety %: clamp the DISPLAY to ±100% so an extreme value (e.g. a method
    whose book value is distorted) doesn't render as a broken-looking huge number. The
    underlying signal still uses the true (unclamped) value, so the color is unaffected.
    """
    if is_missing(v):
        return EM_DASH
    if v >= 100:
        return ">+100%"
    if v <= -100:
        return "<−100%"          # real minus sign (U+2212)
    return f"{v:+.1f}%"


def fmt_delta(yoy_pct: Optional[float]) -> tuple[str, str]:
    """Returns (label_html, css_class) for a YoY delta in the KPI strip."""
    if is_missing(yoy_pct):
        return ("· no prior year", "flat")
    if yoy_pct > 0.05:
        return (f"▲ {yoy_pct:.1f}% YoY", "up")
    if yoy_pct < -0.05:
        return (f"▼ {abs(yoy_pct):.1f}% YoY", "down")
    return (f"≈ {yoy_pct:+.1f}% YoY", "flat")


def fmt_cagr(start, end, years: int) -> tuple[str, str]:
    """Compute CAGR and return (label, css_class). 'n/a' for sign changes or missing endpoints."""
    if is_missing(start) or is_missing(end) or years < 1:
        return (EM_DASH, "")
    if start == 0:
        return ("n/a", "")
    # CAGR is undefined when the series crosses zero (sign change).
    if (start < 0) != (end < 0):
        return ("n/a", "")
    if start < 0 and end < 0:
        # Both negative: compute on absolute values, flip the interpretation.
        ratio = abs(end) / abs(start)
        cagr = (ratio ** (1 / years) - 1) * 100
        cagr = -cagr  # "less negative" = positive trend
    else:
        ratio = end / start
        cagr = (ratio ** (1 / years) - 1) * 100
    label = f"{cagr:+.1f}%"
    cls = "up" if cagr > 0.05 else ("down" if cagr < -0.05 else "")
    return (label, cls)


def trend_growth(values) -> tuple[Optional[float], Optional[float]]:
    """Robust annual growth over the whole series via log-linear regression.
    Fits OLS to ln(value) vs. year offset, annualizes the slope:
        growth% = (exp(slope) - 1) * 100
    Every observation pulls on the fit, so a single anomalous endpoint can't
    swing the read. Returns (growth_pct, r_squared):
      * strictly positive series -> log-linear fit; r_squared is in LOG space
        (how steadily exponential the trend is -- itself a quality signal);
      * all-negative series -> fit on |value|, sign flipped, mirroring fmt_cagr;
      * sign-crossing, or < 3 usable points -> (None, None).
    """
    pts = [(i, v) for i, v in enumerate(values) if not is_missing(v)]
    if len(pts) < 3:
        return (None, None)
    vals = [v for _, v in pts]
    all_pos = all(v > 0 for v in vals)
    all_neg = all(v < 0 for v in vals)
    if not (all_pos or all_neg):
        return (None, None)
    xs = [float(i) for i, _ in pts]
    ys = [math.log(abs(v)) for _, v in pts]
    n = len(xs)
    x_bar, y_bar = sum(xs) / n, sum(ys) / n
    sxx = sum((x - x_bar) ** 2 for x in xs)
    if sxx == 0:
        return (None, None)
    slope = sum((x - x_bar) * (y - y_bar) for x, y in zip(xs, ys)) / sxx
    growth = (math.exp(slope) - 1) * 100.0
    if all_neg:
        growth = -growth
    intercept = y_bar - slope * x_bar
    ss_tot = sum((y - y_bar) ** 2 for y in ys)
    ss_res = sum((y - (slope * x + intercept)) ** 2 for x, y in zip(xs, ys))
    r2 = (1.0 - ss_res / ss_tot) if ss_tot > 0 else None
    return (growth, r2)


def fmt_trend_cagr(values) -> tuple[str, str, str]:
    """Robust trend growth for display. Returns (label, css_class, r2_badge)."""
    growth, r2 = trend_growth(values)
    if growth is None:
        return ("n/a", "", "")
    label = f"{growth:+.1f}%"
    cls = "up" if growth > 0.05 else ("down" if growth < -0.05 else "")
    badge = f"R²&nbsp;{r2:.2f}" if r2 is not None else ""
    return (label, cls, badge)


def short_year(fiscal_year: int) -> str:
    """2024 → '24."""
    return f"'{str(fiscal_year)[-2:]}"


def short_quarter(period_type: str, fiscal_year: int) -> str:
    """('Q4', 2024) → '24-Q4."""
    return f"'{str(fiscal_year)[-2:]}-{period_type}"

File: fundamentals_pipeline/60__frontends/62__web/src/lib/format.ts
// Number formatting — port of the subset of 61__streamlit/lib/format.py used by the company
// masthead, KPI strip and derived-metrics grid. Keep in sync with that module.

export const EM_DASH = "—";

export function isMissing(v: number | null | undefined): v is null | undefined {
  return v === null || v === undefined || Number.isNaN(v);
}

/** KPI strip / usd metric: $391.0B / $93.7M / $1,234 (scale-aware, one decimal). */
export function fmtKpi(v: number | null | undefined): string {
  if (isMissing(v)) return EM_DASH;
  const sign = v < 0 ? "-" : "";
  const a = Math.abs(v);
  if (a >= 1_000_000_000) return `${sign}$${(a / 1_000_000_000).toFixed(1)}B`;
  if (a >= 1_000_000) return `${sign}$${(a / 1_000_000).toFixed(1)}M`;
  return `${sign}$${a.toLocaleString("en-US", { maximumFractionDigits: 0 })}`;
}

/**
 * Derived metric, formatted by unit. `signed` forces a leading "+" on positive percents —
 * only for direction metrics (YoY growth, margin of safety), never levels (margins/returns).
 */
export function fmtMetric(
  v: number | null | undefined,
  unit: string | null | undefined,
  signed = false,
): string {
  if (isMissing(v)) return EM_DASH;
  if (unit === "percent") return signed ? `${v >= 0 ? "+" : ""}${v.toFixed(1)}%` : `${v.toFixed(1)}%`;
  if (unit === "ratio") return `${v.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}x`;
  if (unit === "usd") return fmtKpi(v);
  return v.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

/** Margin-of-safety %: clamp the DISPLAY to ±100% (the signal still uses the true value). */
export function fmtMos(v: number | null | undefined): string {
  if (isMissing(v)) return EM_DASH;
  if (v >= 100) return ">+100%";
  if (v <= -100) return "<−100%"; // real minus sign U+2212
  return `${v >= 0 ? "+" : ""}${v.toFixed(1)}%`;
}

export type DeltaClass = "up" | "down" | "flat";

/** YoY delta for a KPI card: [label, css class]. */
export function fmtDelta(yoyPct: number | null | undefined): [string, DeltaClass] {
  if (isMissing(yoyPct)) return ["· no prior year", "flat"];
  if (yoyPct > 0.05) return [`▲ ${yoyPct.toFixed(1)}% YoY`, "up"];
  if (yoyPct < -0.05) return [`▼ ${Math.abs(yoyPct).toFixed(1)}% YoY`, "down"];
  return [`≈ ${yoyPct >= 0 ? "+" : ""}${yoyPct.toFixed(1)}% YoY`, "flat"];
}

/** (latest − prior) / |prior| as a percent, or null when not computable. Series is oldest→newest. */
export function yoy(values: (number | null)[]): number | null {
  const n = values.length;
  if (n < 2) return null;
  const a = values[n - 1];
  const b = values[n - 2];
  if (isMissing(a) || isMissing(b) || b === 0) return null;
  return ((a - b) / Math.abs(b)) * 100;
}

/** Latest non-missing value of an oldest→newest series, or null. */
export function latestOf(values: (number | null)[]): number | null {
  const last = values[values.length - 1];
  return isMissing(last) ? null : last;
}

const commas = (v: number, dp = 0) =>
  v.toLocaleString("en-US", { minimumFractionDigits: dp, maximumFractionDigits: dp });

/** Statement number cell: thousands-separated integer, accounting-style negatives. */
export function fmtNum(v: number | null | undefined): string {
  if (isMissing(v)) return EM_DASH;
  return v < 0 ? `(${commas(Math.abs(v))})` : commas(v);
}

/** Statement cell with an optional unit scale (1 / 1e3 / 1e6 / 1e9). divisor 1 ⇒ fmtNum. */
export function fmtNumScaled(v: number | null | undefined, divisor = 1): string {
  if (isMissing(v)) return EM_DASH;
  if (divisor === 1) return fmtNum(v);
  const scaled = v / divisor;
  return scaled < 0 ? `(${commas(Math.abs(scaled), 1)})` : commas(scaled, 1);
}

/** Per-share row: 2 decimals, accounting-style negatives. */
export function fmtEps(v: number | null | undefined): string {
  if (isMissing(v)) return EM_DASH;
  return v < 0 ? `(${commas(Math.abs(v), 2)})` : commas(v, 2);
}

/** CAGR over `years` periods → [label, css class]. "n/a" on sign change / zero start. */
export function fmtCagr(
  start: number | null,
  end: number | null,
  years: number,
): [string, "" | "up" | "down"] {
  if (isMissing(start) || isMissing(end) || years < 1) return [EM_DASH, ""];
  if (start === 0) return ["n/a", ""];
  if (start < 0 !== end < 0) return ["n/a", ""]; // crosses zero → undefined
  let cagr: number;
  if (start < 0 && end < 0) {
    cagr = -((Math.abs(end) / Math.abs(start)) ** (1 / years) - 1) * 100;
  } else {
    cagr = ((end / start) ** (1 / years) - 1) * 100;
  }
  const cls = cagr > 0.05 ? "up" : cagr < -0.05 ? "down" : "";
  return [`${cagr >= 0 ? "+" : ""}${cagr.toFixed(1)}%`, cls];
}

/**
 * Robust annual growth via log-linear OLS on ln|value| vs year offset → [growth%, r²].
 * Strictly-positive or all-negative series only; sign-crossing or <3 points → [null, null].
 */
export function trendGrowth(values: (number | null)[]): [number | null, number | null] {
  const pts = values
    .map((v, i) => [i, v] as const)
    .filter(([, v]) => !isMissing(v)) as [number, number][];
  if (pts.length < 3) return [null, null];
  const vals = pts.map(([, v]) => v);
  const allPos = vals.every((v) => v > 0);
  const allNeg = vals.every((v) => v < 0);
  if (!allPos && !allNeg) return [null, null];
  const xs = pts.map(([i]) => i);
  const ys = pts.map(([, v]) => Math.log(Math.abs(v)));
  const n = xs.length;
  const xBar = xs.reduce((a, b) => a + b, 0) / n;
  const yBar = ys.reduce((a, b) => a + b, 0) / n;
  const sxx = xs.reduce((a, x) => a + (x - xBar) ** 2, 0);
  if (sxx === 0) return [null, null];
  const slope = xs.reduce((a, x, i) => a + (x - xBar) * (ys[i] - yBar), 0) / sxx;
  let growth = (Math.exp(slope) - 1) * 100;
  if (allNeg) growth = -growth;
  const intercept = yBar - slope * xBar;
  const ssTot = ys.reduce((a, y) => a + (y - yBar) ** 2, 0);
  const ssRes = xs.reduce((a, x, i) => a + (ys[i] - (slope * x + intercept)) ** 2, 0);
  const r2 = ssTot > 0 ? 1 - ssRes / ssTot : null;
  return [growth, r2];
}

/** Robust trend growth for display → [label, css class, r² badge text]. */
export function fmtTrendCagr(values: (number | null)[]): [string, "" | "up" | "down", string] {
  const [growth, r2] = trendGrowth(values);
  if (growth === null) return ["n/a", "", ""];
  const cls = growth > 0.05 ? "up" : growth < -0.05 ? "down" : "";
  const badge = r2 !== null ? `R² ${r2.toFixed(2)}` : "";
  return [`${growth >= 0 ? "+" : ""}${growth.toFixed(1)}%`, cls, badge];
}

/** 2024 → '24. */
export function shortYearLabel(year: number): string {
  return `'${String(year).slice(-2)}`;
}
